server/src/main/java/org/opensearch/search/querygroup/tracker/QueryGroupResourceUsageTrackerService.java

```diff
@@ -0,0 +1,123 @@
+    }
+
+    /**
+     * Constructs a map of QueryGroupLevelResourceUsageView instances for each QueryGroup.
+     *
+     * @return Map of QueryGroup views
+     */
+    @Override
+    public Map<String, QueryGroupLevelResourceUsageView> constructQueryGroupLevelUsageViews() {
+        Map<String, QueryGroupLevelResourceUsageView> queryGroupViews = new HashMap<>();
+
+        Map<String, List<Task>> tasksByQueryGroup = getTasksGroupedByQueryGroup();
+        Map<String, Map<ResourceType, Long>> queryGroupResourceUsage = getResourceUsageOfQueryGroups(tasksByQueryGroup);
+
+        for (String queryGroupId : tasksByQueryGroup.keySet()) {
+            QueryGroupLevelResourceUsageView queryGroupLevelResourceUsageView = new QueryGroupLevelResourceUsageView(
+                queryGroupId,
+                queryGroupResourceUsage.get(queryGroupId),
+                tasksByQueryGroup.get(queryGroupId)
+            );
+            queryGroupViews.put(queryGroupId, queryGroupLevelResourceUsageView);
+        }
+        return queryGroupViews;
+    }
+
+    /**
+     * Groups tasks by their associated QueryGroup.
+     *
+     * @return Map of tasks grouped by QueryGroup
+     */
+    private Map<String, List<Task>> getTasksGroupedByQueryGroup() {
+        return taskResourceTrackingService.getResourceAwareTasks()
+            .values()
+            .stream()
+            .filter(QueryGroupTask.class::isInstance)
+            .map(QueryGroupTask.class::cast)
+            .collect(Collectors.groupingBy(QueryGroupTask::getQueryGroupId, Collectors.mapping(task -> (Task) task, Collectors.toList())));
+    }
+
+    /**
+     * Calculates the resource usage of each QueryGroup.
+     *
+     * @param tasksByQueryGroup Map of tasks grouped by QueryGroup
+     * @return Map of resource usage for each QueryGroup
+     */
+    private Map<String, Map<ResourceType, Long>> getResourceUsageOfQueryGroups(Map<String, List<Task>> tasksByQueryGroup) {
+        Map<String, Map<ResourceType, Long>> resourceUsageOfQueryGroups = new HashMap<>();
+
+        // Iterate over each QueryGroup entry
+        for (Map.Entry<String, List<Task>> queryGroupEntry : tasksByQueryGroup.entrySet()) {
+            String queryGroupId = queryGroupEntry.getKey();
+            List<Task> tasks = queryGroupEntry.getValue();
+
+            // Prepare a usage map for the current QueryGroup, or retrieve the existing one
+            Map<ResourceType, Long> queryGroupUsage = resourceUsageOfQueryGroups.computeIfAbsent(queryGroupId, k -> new HashMap<>());
+
+            // Accumulate resource usage for each task in the QueryGroup
+            for (Task task : tasks) {
+                for (ResourceType resourceType : TRACKED_RESOURCES) {
+                    long currentUsage = queryGroupUsage.getOrDefault(resourceType, 0L);
+                    long taskUsage = resourceType.getResourceUsage(task);
+                    queryGroupUsage.put(resourceType, currentUsage + taskUsage);
```

Comment:
Instead of making get/put call for every task usage we can update it just once after we have accumulated the value for all the tasks against the query group

```
for (ResourceType resourceType : TRACKED_RESOURCES) {
        long queryGroupResourceUsage = 0;
        for (Task task : tasks) {
             queryGroupResourceUsage += resourceType.getResourceUsage(task);
        }
        queryGroupUsage.put(resourceType, queryGroupResourceUsage);
}
```

Reply:
Considering the tasks count will be way higher compared to the count of the queryGroups. 
I think iterating over tasks once is the most optimal especially considering the cost with updating queryGroups usage is a get/put on HashMap. (cc: @jainankitk )

Reply:
Map's `get` and `put` logic is way more complex and compute heavy than iterating over list of tasks. This change doesn't change anything in terms of repeating things for query group.

the net gain here is per task which can be viewed as simple as
```
costOf(get|put) - iteration of the task from list
```

Snippet of `get` method
```
final Node<K,V> getNode(Object key) {
        Node<K,V>[] tab; Node<K,V> first, e; int n, hash; K k;
        if ((tab = table) != null && (n = tab.length) > 0 &&
            (first = tab[(n - 1) & (hash = hash(key))]) != null) {
            if (first.hash == hash && // always check first node
                ((k = first.key) == key || (key != null && key.equals(k))))
                return first;
            if ((e = first.next) != null) {
                if (first instanceof TreeNode)
                    return ((TreeNode<K,V>)first).getTreeNode(hash, key);
                do {
                    if (e.hash == hash &&
                        ((k = e.key) == key || (key != null && key.equals(k))))
                        return e;
                } while ((e = e.next) != null);
            }
        }
        return null;
    }
```
This will definitely take more CPU cycles then accessing a single entry from the list.

Reply:
Given we are seeing every queryGroupId once, I don't think computeIfAbsent is really needed. I will rewrite this whole method as one of below:

```
    // Prepare a usage map for the current QueryGroup
    Map<String, EnumMap<ResourceType, Long>> resourceUsageOfQueryGroups = tasksByQueryGroup.keySet().stream().collect(Collectors.toMap(Function.identity(), k -> new EnumMap<>(ResourceType.class)));

    // Iterate over each QueryGroup entry
    for (Map.Entry<String, List<Task>> queryGroupEntry : tasksByQueryGroup.entrySet()) {
        EnumMap<ResourceType, Long> queryGroupUsage = resourceUsageOfQueryGroups.get(queryGroupEntry.getKey());
        for (ResourceType resourceType : TRACKED_RESOURCES) {
            long queryGroupResourceUsage = 0;
            for (Task task : queryGroupEntry.getValue()) {
                 queryGroupResourceUsage += resourceType.getResourceUsage(task);
            }
            queryGroupUsage.put(resourceType, queryGroupResourceUsage);
        }
    }
```

OR

```
// Prepare a usage map for the current QueryGroup
    Map<String, EnumMap<ResourceType, Long>> resourceUsageOfQueryGroups = tasksByQueryGroup.keySet().stream().collect(Collectors.toMap(Function.identity(), k -> new EnumMap<>(ResourceType.class)));

    // Iterate over each QueryGroup entry
    for (Map.Entry<String, List<Task>> queryGroupEntry : tasksByQueryGroup.entrySet()) {
        final String queryGroupId = queryGroupEntry.getKey();
        EnumMap<ResourceType, Long> queryGroupUsage = resourceUsageOfQueryGroups.get(queryGroupId);
        for (ResourceType resourceType : TRACKED_RESOURCES) {
            queryGroupUsage.put(resourceType, 0);
            for (Task task : queryGroupEntry.getValue()) {
                 queryGroupUsage.put(queryGroupId, (k,v) -> v + resourceType.getResourceUsage(task));
            }
        }
    }
```

**Note:** Both look good, since we are using EnumMap (internally represented as arrays) instead of HashMap. Else, I would have preferred the first variant to avoid multiple put calls!

Reply:
very interesting, I didn't think this way, i have made the changes, thanks !